### PR TITLE
Make sure the desired XR has type metadata 

### PR DESCRIPTION
**Pull request by negz**
This is required for it to survive a round-trip through JSON marshalling and unmarshalling, which patching depends on. If it doesn't have a kind field, the Unstructured type's UnmarshalJSON method will return an error.

Fixes https://github.com/crossplane-contrib/function-patch-and-transform/issues/8

### CODE REVIEW

File: patches.go [hunk 1]
```diff
@@ -34,20 +34,20 @@ func Apply(p v1beta1.Patch, cp resource.Composite, cd resource.Composed, only ..
...
+		return ApplyFromFieldPathPatch(p, xr, cd)
 	case v1beta1.PatchTypeToCompositeFieldPath:
-		return ApplyFromFieldPathPatch(p, cd, cp)
+		return ApplyFromFieldPathPatch(p, cd, xr)
 	case v1beta1.PatchTypeCombineFromComposite:
-		return ApplyCombineFromVariablesPatch(p, cp, cd)
+		return ApplyCombineFromVariablesPatch(p, xr, cd)
 	case v1beta1.PatchTypeCombineToComposite:
-		return ApplyCombineFromVariablesPatch(p, cd, cp)
+		return ApplyCombineFromVariablesPatch(p, cd, xr)
```
Comment (tnthornton) [inline patches.go:50]: I appreciate this 😄. There's been a good chunk of time I've lost to trying to reason out the subtle cp v cd.

File: patches_test.go [hunk 2]
```diff
@@ -818,6 +731,14 @@ func TestPatchApply(t *testing.T) {
 	}
 }
 
+func MustObject(j string) map[string]any {
```
Comment (tnthornton) [inline patches_test.go:734]: Is this worth promoting to c/crossplane-runtime at some point in the future?

**Review by tnthornton — APPROVED** [review]
Thanks @negz !

**Comment by negz** [comment]
If I run `xrender` against 2ce28cba6d1510c47811ac6fbfec070e3df6f56a I get:

```
$ git co main
$ go run . --insecure -d
2023-09-29T15:55:15.898-0700    INFO    function-patch-and-transform/fn.go:29   Running Function        {"tag": ""}
2023-09-29T15:55:15.899-0700    DEBUG   function-patch-and-transform/fn.go:96   Processing resource template    {"xr-version": "aws.caas.upbound.io/v1alpha1", "xr-kind": "Cluster", "xr-name": "aws-spoke-01", "resource-template-name": "composite-network-eks"}
2023-09-29T15:55:15.899-0700    DEBUG   function-patch-and-transform/fn.go:153  Found corresponding observed resource   {"xr-version": "aws.caas.upbound.io/v1alpha1", "xr-kind": "Cluster", "xr-name": "aws-spoke-01", "resource-template-name": "composite-network-eks", "ready": false, "name": "existing-xnetwork"}
2023-09-29T15:55:15.899-0700    INFO    function-patch-and-transform/fn.go:170  Cannot render ToComposite patches for composed resource {"xr-version": "aws.caas.upbound.io/v1alpha1", "xr-kind": "Cluster", "xr-name": "aws-spoke-01", "resource-template-name": "composite-network-eks", "warning": "cannot apply the \"ToCompositeFieldPath\" patch at index 5: Object 'Kind' is missing in '{\"status\":{\"subnetIds\":[\"subnet-a\",\"subnet-b\"]}}'"}
2023-09-29T15:55:15.899-0700    DEBUG   function-patch-and-transform/fn.go:96   Processing resource template    {"xr-version": "aws.caas.upbound.io/v1alpha1", "xr-kind": "Cluster", "xr-name": "aws-spoke-01", "resource-template-name": "composite-cluster-eks"}
2023-09-29T15:55:15.899-0700    INFO    function-patch-and-transform/fn.go:182  Cannot render FromComposite patches for composed resource       {"xr-version": "aws.caas.upbound.io/v1alpha1", "xr-kind": "Cluster", "xr-name": "aws-spoke-01", "resource-template-name": "composite-cluster-eks", "warning": "cannot apply the \"FromCompositeFieldPath\" patch at index 12: status: no such field"}
2023-09-29T15:55:15.899-0700    DEBUG   function-patch-and-transform/fn.go:96   Processing resource template    {"xr-version": "aws.caas.upbound.io/v1alpha1", "xr-kind": "Cluster", "xr-name": "aws-spoke-01", "resource-template-name": "composite-cluster-services"}
2023-09-29T15:55:15.899-0700    INFO    function-patch-and-transform/fn.go:201  Successfully processed patch-and-transform resources    {"xr-version": "aws.caas.upbound.io/v1alpha1", "xr-kind": "Cluster", "xr-name": "aws-spoke-01", "resource-templates": 3, "existing-resources": 1, "warnings": 2}
```

Whereas after this change I get:

```
$ go run . --insecure -d
2023-09-29T15:57:00.271-0700    INFO    function-patch-and-transform/fn.go:31   Running Function        {"tag": ""}
2023-09-29T15:57:00.272-0700    DEBUG   function-patch-and-transform/fn.go:110  Processing resource template    {"tag": "", "xr-version": "aws.caas.upbound.io/v1alpha1", "xr-kind": "Cluster", "xr-name": "aws-spoke-01", "resource-template-name": "composite-network-eks"}
2023-09-29T15:57:00.272-0700    DEBUG   function-patch-and-transform/fn.go:138  Resource template corresponds to existing composed resource     {"tag": "", "xr-version": "aws.caas.upbound.io/v1alpha1", "xr-kind": "Cluster", "xr-name": "aws-spoke-01", "resource-template-name": "composite-network-eks", "metadata-name": "existing-xnetwork"}
2023-09-29T15:57:00.272-0700    DEBUG   function-patch-and-transform/fn.go:167  Found corresponding observed resource   {"tag": "", "xr-version": "aws.caas.upbound.io/v1alpha1", "xr-kind": "Cluster", "xr-name": "aws-spoke-01", "resource-template-name": "composite-network-eks", "ready": false, "name": "existing-xnetwork"}
2023-09-29T15:57:00.273-0700    DEBUG   function-patch-and-transform/fn.go:110  Processing resource template    {"tag": "", "xr-version": "aws.caas.upbound.io/v1alpha1", "xr-kind": "Cluster", "xr-name": "aws-spoke-01", "resource-template-name": "composite-cluster-eks"}
2023-09-29T15:57:00.273-0700    INFO    function-patch-and-transform/fn.go:196  Cannot render FromComposite patches for composed resource       {"tag": "", "xr-version": "aws.caas.upbound.io/v1alpha1", "xr-kind": "Cluster", "xr-name": "aws-spoke-01", "resource-template-name": "composite-cluster-eks", "warning": "cannot apply the \"FromCompositeFieldPath\" patch at index 12: status: no such field"}
2023-09-29T15:57:00.273-0700    DEBUG   function-patch-and-transform/fn.go:110  Processing resource template    {"tag": "", "xr-version": "aws.caas.upbound.io/v1alpha1", "xr-kind": "Cluster", "xr-name": "aws-spoke-01", "resource-template-name": "composite-cluster-services"}
2023-09-29T15:57:00.273-0700    INFO    function-patch-and-transform/fn.go:215  Successfully processed patch-and-transform resources    {"tag": "", "xr-version": "aws.caas.upbound.io/v1alpha1", "xr-kind": "Cluster", "xr-name": "aws-spoke-01", "resource-templates": 3, "existing-resources": 1, "warnings": 1}
```

File: patches.go (hunk 1 above)
Comment (negz) [reply patches.go:50]: Yeah, I think the `cp` for "composite" actually predates the introduction of the term "XR". Now that XR is prevalent it definitely makes more sense as a variable name.

File: patches_test.go (hunk 2 above)
Comment (negz) [reply patches_test.go:734]: I think so - I've definitely written it a few times now. I don't think it's super urgent though given how tiny it is.

**Pull request merged**